x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java

```diff
@@ -0,0 +1,108 @@
+                return;
+            }
+            if (httpKeyConfig instanceof StoreKeyConfig == false) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration is not configured with a keystore"));
+                return;
+            }
+            final List<Tuple<PrivateKey, X509Certificate>> transportKeysAndCertificates =
+                ((StoreKeyConfig) transportKeyConfig).getPrivateKeyEntries(environment);
+            final List<Tuple<PrivateKey, X509Certificate>> httpCaKeysAndCertificates =
+                ((StoreKeyConfig) httpKeyConfig).getPrivateKeyEntries(environment).stream()
+                    .filter(t -> t.v2().getBasicConstraints() != -1).collect(Collectors.toList());
+            if (transportKeysAndCertificates.isEmpty()) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll node. Elasticsearch node transport layer SSL configuration doesn't contain any keys"));
+                return;
+            } else if (transportKeysAndCertificates.size() > 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll node. Elasticsearch node transport layer SSL configuration contains multiple keys"));
+                return;
+            }
+            if (httpCaKeysAndCertificates.isEmpty()) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain any " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            } else if (httpCaKeysAndCertificates.size() > 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to enroll client. Elasticsearch node HTTP layer SSL configuration Keystore contain multiple " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            }
+
+            final String httpCaKey = Base64.getUrlEncoder().encodeToString(httpCaKeysAndCertificates.get(0).v1().getEncoded());
+            final String httpCaCert = Base64.getUrlEncoder().encodeToString(httpCaKeysAndCertificates.get(0).v2().getEncoded());
+            final String transportKey = Base64.getUrlEncoder().encodeToString(transportKeysAndCertificates.get(0).v1().getEncoded());
+            final String transportCert = Base64.getUrlEncoder().encodeToString(transportKeysAndCertificates.get(0).v2().getEncoded());
+            final List<String> nodeList = new ArrayList<>();
+            for (DiscoveryNode node : clusterService.state().getNodes()) {
+                nodeList.add(node.getAddress().toString());
+            }
```

Comment:
What do you think if we use the output of `NodeEnrollmentAction.NAME` with the transport metric parameter?
It gives me more confidence that we're returning the right addresses, and the client might already be used to parsing that.

Reply:
I'm not sure I understood the suggestion @albertzaharovits .You mean that we should return `enroll:enroll_node:192.168.1.2:9300` instead of `192.168.1.2:9300` ? What does this buy us ?

Reply:
Bleah... sorry, I messed up the clipboard buffers.
I meant `NodeEnrollmentAction.NAME` -> `NodesInfoAction.NAME` .

Reply:
makes sense, I'm calling this for the enroll client API transport action, might as well do this here too, no objections